stack implementation and tests

LeonidGoltsblat opened the pull request:
This is a stack container implementation for the pjsib library (pj_stack), i.e. the single-linked list with First In Last Out logic (FILO). 
The stack is implemented as close as possible to the implementation of pj_list.
This implementation is thread safe. Common implementation uses internal locking mechanism so is thread-safe.
But implementation for Windows platform uses locking free Windows embeded single linked list implementation, which makes this implementation exceptionally fast.

This is one of the basic mechanisms used in a series of subsequent pull requests with the overall goal of improving pjsip's performance.

The pull request contains tests that can be used as usage examples.

pjlib/src/pj/stack.c is the common stack implementation
pjlib/src/pj/stack_win32.c is the implementation for Windows platform

see more info in embedded documentation

sauwming commented:
Thank you for the contribution. Perhaps you can tell us the reason of the introduction of this new data structure? (i.e. what future PR/feature do you plan to submit that will require the usage of stack?)

LeonidGoltsblat commented:
Sorry for the delay in replying, I will definitely reply, but I will be busy for a couple more days.

LeonidGoltsblat commented:
Sorry for the delay in replying,
Yes, I'm working on a few poll requests..
The main motivation for all of them is improved pjsip performance when servicing many calls at once
- iocp (ioqueue_winnt). Some tests are now showing 10*n times perfomance enhanced compare to standard ioqueue_select
- optimized udp_rtp transport (for example media stream changes implemented as it's closing and reopen should not force waiting states for other rtp stream)
- parallelized conference bridge (yes, I see new commits with "async conference bridge", maybe I'm late...)
- concept of "external data source" for "file player" - API for reading playback data from virtually any source (for example from a BLOB field in a database). As a special case, I implemented asynchronous pre-emptive file reading for use in IVR scenarios with many simultaneous file playbacks
- some "small" changes, like ability to track lock wait duration and lock hold duration, and changes in pjsip algorithms to minimize the duration of such wait states. 
I often use "stack" implementation as a very quick alternative to "list", espessialy agressively in the previous point.

sauwming commented:
No worries at all about the delay, we are preparing for the final testing of 2.15 release so most likely, we can't merge this until the release anyway.

It's still not obvious to me from your examples when you need to use the stack? Typically in the SIP context, we use queue since we need to process the messages/events that come first (FIFO). In what specific case would you use LIFO?

LeonidGoltsblat commented:
I added stack_stress_test() as a one of the typical examples when we can use the stack (reserving an empty slot in a large array without having to lock the entire array).

sauwming commented:
Quite frankly, the modifications are much more comprehensive than I expected.

There doesn't seem to be any strong reason for a platform-specific stack implementation (i.e. stack_win32), especially since it introduces more complexities and restrictions, such as the memory alignment requirements. Portability and simplicity should be the objective here. For example, we even introduced the use of linked list for our timer implementation (which is O(n)), instead of heap data structure (which is O(log n)), simply because the former is much simpler and easier to debug.

So I propose to keep the patch simple, one stack implementation for all platforms, and one simple unit test should suffice (similar to unit test for list, no need for stress test since we should not be aiming for high performance data structure).

LeonidGoltsblat commented:
You are quite right, it became too comprehensive!
The only reason for this is that I have no access to test environment config_side.h file (isn't it? May be I don't know the way?)
My initial idea was to have something like this in my config_site.h
#ifdef PJ_WIN32
#if !defined(MEMORY_ALLOCATION_ALIGNMENT)
#include <windows.h>
#endif
//our pj_stack implementation requires such alignment
#define PJ_POOL_ALIGNMENT    MEMORY_ALLOCATION_ALIGNMENT
#endif
This eliminates the needness of other alignment related changes I made.
With this line in config_site.h I may check data alignment in PJ_ASSERT expression and even in compile time.
All of this still exists in stack_win32.c.
if pool return properly alignment memory, setting alignment explicitly for each declared stack item type is optional (but still required if we have an array whose items should be the stack items).

User still may to use "crossplatform" (but ineffective under Windows) implementation without setting PJ_POOL_ALIGNMENT in config_site.h (as it is now for iocp and select io_queue)

If it is possible to add lines above into config_site.h, everything become simple! What do you think about this option?

PS For me, performance under Windows is critical.

LeonidGoltsblat commented:
Ok. I pushed a simplified cross-platform implementation. Now by default all platforms use the same implementation, but some weird users (like me...) who need extra performance on Windows can add
#define PJ_USE_CROSSPLATFORM_STACK_IMPL 0
and something like
#define PJ_POOL_ALIGNMENT MEMORY_ALLOCATION_ALIGNMENT
to their config_site.h
In this case, the Windows platform implementation will be compiled

nanangizz commented:
The idea of improving performance in thread-safe environment (via lock-free method & data alignment) is interesting, but perhaps it is more effective if the features (lock-free & data alignment) are added to the existing PJLIB list, a generic type and used heavily in the library (and stack may be implemented as a lite wrapper on top of it), and, as @sauwming mentioned, is targeting more platforms (especially major ones: Apple, Linux/*nix/Android, and Windows).

Re: thread-safety, as you may know already, currently PJLIB collection types (e.g: list, array) are not thread-safe by design and this is okay because in many cases applications need to synchronize more than one states at once, not just a state (e.g: list). So the synchronization is to be done in the app, and using a thread-safe list may even be a bit disadvantage as it may add a slight unnecessary overhead. So IMO, thread-safety feature should be configurable.

Also, there seems some irreleveant changes in the patch?

LeonidGoltsblat commented:
@sauwming Thank you for feadback, but I don't fully understand now what I shell to do...
> Quite frankly, the modifications are much more comprehensive than I expected.
> 
> There doesn't seem to be any strong reason for a platform-specific stack implementation (i.e. stack_win32), especially since it introduces more complexities and restrictions, such as the memory alignment requirements. 

> So I propose to keep the patch simple, one stack implementation for all platforms, and one simple unit test should suffice (similar to unit test for list, no need for stress test since we should not be aiming for high performance data structure).

The last change I made was my response to your comment above. Now all platforms default to the same "cross-platform" implementation you asked for earlier. 
To integrate my further work into pjsip, I need an implementation not only for Windows (and not only for VS), but for all platforms, and this implementation must be thread-safe.
If you don't mind a special implementation for Windows now (and setting a special alignment for this platform), I can change the configuration again so that stack_win32 is compiled by default. If so, I can put these macro declarations in config.h. @sauwming What do you think about this?

Furthermore, all the recent changes to the other platforms' build files (and in pjlib/include/pj/compat) are that I removed unnecessary changes I made earlier and added stack.o to just one Makefile. The only change in pjlib/include/pj/compat is a slightly more readable but fully compatible redefinition of the alignment macros. These new macros are used in stack.h.

LeonidGoltsblat commented:
@nanangizz You are absolutely right, a stack can of course be implemented as a list wrapper. But in my use case FILO is not the key functionality. I use pj_stack where I want to get an arbitrary element of a collection, but in a fast and thread-safe way. I wouldn't want to go that direction, but this object could be renamed to hide the FILO implementation.
Of course, we can add a "policy" to this object that will make it thread-safe or not, but is it really necessary? If we want an unsafe stack, we can use pj_list directly, without any wrapper (or we may even add push/pop methods into pj_list).

> Also, there seems some irreleveant changes in the patch?

Maybe, but please explain what changes exactly you mean?

sauwming commented:
> The last change I made was my response to your comment above. Now all platforms default to the same "cross-platform" implementation you asked for earlier. To integrate my further work into pjsip, I need an implementation not only for Windows (and not only for VS), but for all platforms, and this implementation must be thread-safe. If you don't mind a special implementation for Windows now (and setting a special alignment for this platform), I can change the configuration again so that stack_win32 is compiled by default. If so, I can put these macro declarations in config.h. @sauwming What do you think about this?

After reading the overwhelmingly significant performance boost of Windows SList vs regular mutexed stack:
https://kukuruku.co/hub/cpp/lock-free-stack-for-windows
I now understand why you push for the special Windows stack implementation and why it has to be a LIFO stack (instead of a regular list or a FIFO queue).

So:
- Yes, I am okay with special Windows implementation.
- Unfortunately, since no one (except you :) ) is using the stack right now, and using it requires setting the memory alignment for the entire PJSIP library, by default it needs to be disabled (for now). For Windows users that want to use SList, enabling it should be as simple as adding this in `config_site.h`: `#define PJ_STACK_USE_WIN_SLIST 1`. You can set the memory alignment in `pjlib/config.h`.

sauwming commented on pjlib/include/pj/stack.h:
```diff
@@ -0,0 +1,185 @@
+ * 0 - use the Windows platform stack implementation
+ * 1 - force the crossplatform stack implementation.
+ */
+#ifndef PJ_USE_CROSSPLATFORM_STACK_IMPL
```
I just realised that my suggestion is not PJSIP's standard.
Usually in PJSIP we use something like `PJ_STACK_IMPLEMENTATION==PJ_STACK_WIN32`, similar to `sock_qos` or `ssl_sock`.

LeonidGoltsblat replied:
Great idea! Done

nanangizz commented:
> But in my use case FILO is not the key functionality. I use pj_stack where I want to get an arbitrary element of a collection, but in a fast and thread-safe way. I wouldn't want to go that direction, but this object could be renamed to hide the FILO implementation.

This sounds like: instead of stack, this can be implemented a list. And as mentioned before, implementing it as a list, which is heavily used by the library, would be more effective to the idea of improving library performance.

It may be an alternative implementation of the existing list (using the same APIs as much as possible), like IOQueue and TLS, which have some alternative implementations.

In the future perhaps we can add lock-free list implementations for non-Windows.


> Of course, we can add a "policy" to this object that will make it thread-safe or not, but is it really necessary?

As mentioned before, there will be slight overhead if app also manages the synchronization, especially for non-Windows?


> > Also, there seems some irreleveant changes in the patch?
> 
> Maybe, but please explain what changes exactly you mean?

@sauwming seems to have done it :)

nanangizz commented:
>     * Unfortunately, since no one (except you :) ) is using the stack right now, and using it requires ...

IMO this is something we need to consider, once merged, we will have to maintain it (fixing any bugs, etc) while currently nothing is using it :)

LeonidGoltsblat commented on pjlib/include/pj/compat/os_win32.h:
```diff
@@ -148,6 +148,5 @@
 #   endif
 #endif
 
-
```
Ok, file restored

LeonidGoltsblat commented on pjlib/include/pj/compat/os_winuwp.h:
```diff
@@ -145,5 +145,4 @@
  */
 #define PJ_THREAD_ALLOCATE_STACK        0
 
-
```
Ok, file restored

LeonidGoltsblat commented on pjsip-apps/build/pjsystest.vcxproj:
```diff
@@ -1,205 +1,356 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
```
I don't see any changes in the first line of this project file, but it does add the appropriate project configurations to build these tests with the debug-dynamic or release-dynamic libraries.

sauwming replied:
pjsystest (which is for sound device test) should not be affected by the addition of stack, so you should exclude it from the patch.

If you find missing build configurations, perhaps use a separate PR.

LeonidGoltsblat replied:
create PR #4212
but there are no nothing about stack
we will need to merge carefully

pjsystest.vcxproj may be excluded here

LeonidGoltsblat commented:
@nanangizz 

The idea to implement FIFO in a lock-free way is certainly good. I think that implementing full list functionality this way is very difficult and may not be possible at all. I'm not sure, but I think limited functionality like push_back/pop_first can be implemented this way (and that's 99.99...% of the FIFO use cases.)
But I am sure this is not for the current PR... 

>  we will have to maintain it (fixing any bugs, etc)

I'm not sure if it's possible to implement such a FIFO as a hobby, but I'm willing to maintain my code.

> while currently nothing is using it 

when this patch will be aproved I'll try quickly fix this in the next PRs :)

LeonidGoltsblat commented on build/vs/pjproject-vs14-common-config.props:
```diff
@@ -18,20 +18,20 @@
   <PropertyGroup>    
     <API_Family Condition="'$(API_Family)'==''">WinDesktop</API_Family>
     <PreprocessorDef></PreprocessorDef>
-    <DefaultToolset>v140</DefaultToolset>
```
Ok, file restored

sauwming replied:
Still the same. I think you can safely exclude it from the patch.

LeonidGoltsblat replied:
created PR #4212
Now github sees only some changes in this file, but not in other files...
And Fork (the git client I use) shows only changed lines.
The only idea I have is that VS likes Windows CR/LF line endings and can automatically switch to them.

sauwming replied:
Your editorconfig is probably the reason for all these changes.

I understand you may prefer your local settings, but since it's a public project, it's better to use the default config so everybody adopts the same setting.

LeonidGoltsblat commented:
@sauwming @nanangizz !
"CI Linux / ubuntu-video-openh264-1 (pull_request)" DNS test failed with: server error parsing packet: [pj_status_t=320043] Invalid DNS name pointer (PJLIB_UTIL_EDNSINNAMEPTR)

This is exactly unrelated with my changes. Have no idea how to fix that.

sauwming commented on pjlib/build/pjlib.vcxproj.filters:
```diff
@@ -1,449 +1,461 @@
+    <ClCompile Include="..\src\pj\ssl_sock_imp_common.c">
+      <Filter>Source Files</Filter>
+    </ClCompile>
+    <ClCompile Include="..\src\pj\stack.c">
```
It would be better if the patch doesn't modify the entire file, and only add the necessary ones.
Not sure how to do it though.

Same goes for the other `*.filters` file, and `pjlib_test.vcxproj` (strangely, `pjlib.vcxproj looks okay).

nanangizz commented:
> The idea to implement FIFO in a lock-free way is certainly good. I think that implementing full list functionality this way is very difficult and may not be possible at all. I'm not sure, but I think limited functionality like push_back/pop_first can be implemented this way (and that's 99.99...% of the FIFO use cases.)

Honestly I didn't know that, not familiar with lock-free synchronization.

Just did a random search, I found [liblfds](https://liblfds.org), from a quick check on header file `lfds711_list_addonly_singlylinked_ordered.h`, it has 'insert after', 'insert at' (if those are what you mean by full functionality), haven't checked further if those ops are really lock-free.

Btw, a limited functionality with 99.99% use cases, still sounds like an improvement.

> But I am sure this is not for the current PR...

Sure. Just trying to grasp how the current PR (or any other PRs/contributions for which we are very thankful) can improve the library. It is a new & unused type and offering performance via lock-free & data alignment (CMIIW).

Re: new type, we've discussed above. Re: unused, you've promised things we barely know :)

Re: data alignment sounds great, perhaps it can improve all other types as well?

Re: lock-free or thread-safety. It is perhaps nice, but IMO not so urgent for now. If you search list operation (e.g: `pj_list_push_back()`) in the library source for example and check mutex protection around it, it is actually very rare that the mutex protects only an operation on a list, usually it protects other states or multiple list operations (e.g: is_empty() & push_back()). Perhaps that's why so far we've never considered adding thread-safety feature to the list (IIRC, nor heard suggestion on it). Even it may introduce unnecessary overhead which eventually degrades performance.

sauwming commented on pjlib/src/pj/stack_win32.c:
```diff
@@ -0,0 +1,142 @@
+#include <pj/os.h>
```
Could you please add copyright and licensing info in the beginning of the files? (`stack.*` files)
You can add your name if you wish so.
(For an example, refer to the other PJSIP source files).

LeonidGoltsblat replied:
copyright adde

sauwming commented on pjlib/src/pj/stack_win32.c:
```diff
@@ -0,0 +1,142 @@
+ */
+
+#ifndef PJ_WIN32
+#error stack_win32.c should be compiled �nly for Windows platform
```
s/înly/only

LeonidGoltsblat replied:
fixed

sauwming reviewed (commented):
Just need to clean up the project files and you should be mostly good to go.

For an example of a clean project patch, you can check PR 4132 (https://github.com/pjsip/pjproject/pull/4132/files#diff-0c444d946963ac7a6c002817133fdab6c0cef69c43bd85eff0dcd9c6419ed7ad)

LeonidGoltsblat commented:
> Just need to clean up the project files and you should be mostly good to go.

cleaned! i.e. line ending fixed
some "some irreleveant changes in the patch" moved to #4212

LeonidGoltsblat commented:
@nanangizz 

>Just did a random search, I found [liblfds](https://liblfds.org/), from a quick check on header file lfds711_list_addonly_singlylinked_ordered.h, it has 'insert after', 'insert at' (if those are what you mean by full functionality), haven't checked further if those ops are really lock-free.

Anything is possible, but when I said complexity, I meant this: if we insert "new" before or after "old", we have to make sure that "old" still exists in the container. I think this requires some external synchronization, and if so, I think we have to use this synchronization with all other operations.

> Re: lock-free or thread-safety. It is perhaps nice, but IMO not so urgent for now. If you search list operation (e.g: pj_list_push_back()) in the library source for example and check mutex protection around it, it is actually very rare that the mutex protects only an operation on a list, usually it protects other states or multiple list operations (e.g: is_empty() & push_back()). Perhaps that's why so far we've never considered adding thread-safety feature to the list (IIRC, nor heard suggestion on it). Even it may introduce unnecessary overhead which eventually degrades performance.

Yes, there are many such examples but there are also opposite situations showing that external synchronization is not always optimal.

A little search in the single file pjsu_core.c (only because I have fixed "Merged request detected" in this file some days ago)

pjsua_transport_create()

    PJSUA_LOCK();
    /* Find empty transport slot */
    /* Create the transport */
    /* Save the transport */
on_return:
    PJSUA_UNLOCK();

We have an unnecessary long lock, but with pj_stack this can be done it like this:
    /* Quickly! RESERVE empty transport slot form the stack of empty slots */
    /* Create the transport */
    PJSUA_LOCK();
    /* Save the transport */
    PJSUA_UNLOCK();
Here we have very short locking time.

I don't think we often call this function concurrently :) but pjsip uses such an algorithm everywhere and in some cases this global locking may lead to serious perfomance degradation.

another example from the same file

mod_pjsua_on_rx_request()
    PJSUA_LOCK();
    if (rdata->msg_info.msg->line.req.method.id == PJSIP_INVITE_METHOD) {
        processed = pjsua_call_on_incoming(rdata);
    }
    PJSUA_UNLOCK();
Do we really need to protect rdata->msg_info.msg->line.req.method.id under the global lock? May be better so:
    if (rdata->msg_info.msg->line.req.method.id == PJSIP_INVITE_METHOD) {
       PJSUA_LOCK();
        processed = pjsua_call_on_incoming(rdata);
       PJSUA_UNLOCK();
    }
We have a lot of modules registered, each of which gets a global lock and then in most cases realizes that it wasn't the packet for it. I don't have measurements, but I think we may have a real performance penalty here.

These examples just show that external locking is not always a good choice, and internal locking is not always a bad choice.

sauwming commented on pjlib/include/pj/stack.h:
```diff
@@ -0,0 +1,189 @@
+ * Stack in PJLIB is single-linked list with First In Last Out logic. 
+ * Stack is thread safe. Common PJLIB stack implementation uses internal locking mechanism so is thread-safe.
+ * Implementation for Windows platform uses locking free Windows embeded single linked list implementation.
+ * The performance of pj_stack implementation for Windows platform is 2-5x higher than cross-platform.
```
Putting numbers here without test data may not be wise.
So may be just put "considerably" higher/faster.

sauwming commented on pjlib/src/pj/stack.c:
```diff
@@ -0,0 +1,178 @@
+ * because the item count can be changed at any time by another thread.
+ * For Windows platform returns the number of entries in the stack modulo 65535. For example,
+ * if the specified stack contains 65536 entries, pj_stack_size returns zero.
+ *
```
Remove the Windows comment and @param and @return. Unnecessary here.

sauwming commented on pjlib/src/pj/stack_win32.c:
```diff
@@ -0,0 +1,160 @@
+ * because the item count can be changed at any time by another thread.
+ * For Windows platform returns the number of entries in the stack modulo 65535. For example,
+ * if the specified stack contains 65536 entries, pj_stack_size returns zero.
+ *
```
Remove @param and @return doc.

sauwming commented on pjlib/src/pjlib-test/stack.c:
```diff
@@ -0,0 +1,389 @@
+
+    for (i = 0; !rc && i < PJ_ARRAY_SIZE(tests); ++i) {
+        tests[i].state.pool = pool;
+        rc = stack_stress_test(&tests[i]);
```
perhaps we should by default disable the stress test for non-Windows platform (create a macro such as #HAS_STRESS_TEST PJ_WIN32).

It will only burden the CI machine and there's little point in stress testing a relatively slower implementation of LIFO list + mutex.

LeonidGoltsblat replied:
I think we need multithreaded testing for thread safe api, but currently the default implementation is cross platform so may be just Windows testing  is enough.
Create a macro HAS_MT_STACK_STRESS_TEST

all fixed!

> These should be all from me

I hope this will be the case this time! :)

sauwming reviewed (commented):
These should be all from me.

nanangizz commented:
> These examples just show that external locking is not always a good choice, and internal locking is not always a bad choice.

True, of course. I guess this may also answer your previous question?
`Of course, we can add a "policy" to this object that will make it thread-safe or not, but is it really necessary?`

LeonidGoltsblat commented:
@nanangizz
FYI: [Doubly Linked Lists](https://learn.microsoft.com/en-us/windows-hardware/drivers/kernel/singly-and-doubly-linked-lists#:~:text=Linked%20Lists.-,Doubly%20Linked%20Lists,-The%20operating%20system)
This is WDM, not regular WinApi, and only for Windows, but may be your desire to have a lock-free FIFO can be realized!

LeonidGoltsblat commented on pjlib/include/pj/stack.h:
```diff
@@ -0,0 +1,189 @@
+
+#endif // PJ_STACK_IMPLEMENTATION
+
+#include <pj/stack.h>
```
mistake! Thanks!

LeonidGoltsblat commented:
Hi! 
New conflicts resolved. How about approving this PR?

LeonidGoltsblat commented:
Hello, colleagues!
Please see the first pj_stack usage in #4230!

About pj_stack...
Using C11's built in atomic support, it's easy to implement a cross-platform lock-free LIFO (and maybe FIFO too). What do you think of C11?

bennylp commented:
Hi @LeonidGoltsblat, first of all thanks for your contribution, and your time for submitting this. We're just back from (pjsip) office holiday, so coding is definitely slow this time of year. :)

I've read your submission, and here are my comments. As others have said, my first reaction is, why do we need this? What problem does it solve, or what enhancement does it offer? I don't see your patch is addressing any of these two questions quantitatively. For example, if it proposes significant speed improvements, I would like to be convinced with the numbers. First the theoretical performance improvement (e.g. how long to `push()`/`pop()` 1 billion elements), and maybe the projected actual improvement. In the end, it probably won't matter to have 5x speed improvement, if the improvement is only 50 nanoseconds and the whole operation (such as adding/removing ioqueue key) takes 10 msec.

On the other hand, this is quite a significant submission, as it modifies some of our core header files, adds third party copyright (i.e you), and imposes certain alignment on all pool memory allocations. I would probably only check the alignment in the `stack_win32.c`, to avoid imposing changes on the global space. The naming between `pj_stack_type` and `pj_stack_t` is also confusing, I would probably use `pj_stack_node_t` for the node. Each of this is probably solvable, but again, we have to go back to the fundamental question, i.e. why do we need this. Also this only works on Windows.

So I tend to say we keep this aside for now until we can be convinced there is a real, significant usage for it.

LeonidGoltsblat commented:
Hi Benny, Merry Christmas and thanks for the feedback! I'm currently preparing a PR with a parallel version of the conference bridge. Hopefully it'll only take a few days, then I'll be back to discuss and happy to answer any questions you may have.
Best regards, Leonid

bennylp commented:
The parallel conf bridge sounds exciting! But if the purpose is to show another sample implementation where the lock free stack can be used, I'm afraid this is sounding more and more like the stack is a solution looking for a problem :)

LeonidGoltsblat commented:
@bennylp Please see #4241
By default, the stack is not used there (unfortunately) 🙂.

sauwming commented:
Just an idea, perhaps it's better renaming the data structure from stack to something like `atomic_slist` (to represent an atomic/interlocked singly linked list).

The reason is simple. Using stack to solve the various problems in the other PR seems kind of strange, and as @bennylp pointed out, sounds forced as if the stack is made to solve something it's not supposed to.

But with the more appropriate name, suddenly it just seems natural. To solve the problem of resource allocation (such as unused slots) stored in an array that requires synchronisation, it makes sense IMO to use an `atomic_slist` data structure.

And the data structure also opens the door to be used elsewhere in the library that currently uses the doubly-linked list declared in `list.h` but which can actually be changed into a singly linked version.

bennylp commented:
I'm closing this for now, with the following suggestion if similar work is to be resubmitted in the future:

- please provide justification of the feature (bug fix? speed improvement results?)
- use (currently hypothetical :) `pj_pool_aligned_alloc()` to localize memory alignment requirement to specific data rather than all pool allocations
- use naming like `pj_atomic_slist` to make it more natural.

LeonidGoltsblat commented:
@bennylp 

- Can I expect the hypothetical pj_pool_aligned_alloc() function to become real?
- atomic_slist (and potentially FIFO) can be implemented as thread-safe and cross-platform using C11's atomic support. What are your thoughts on adopting C11 now, in 2025?

bennylp commented:
@LeonidGoltsblat

- yes
- we've discussed this internally as recent as couple of weeks ago. 

  The problem is, unlike C++ standards, newer C standards are not well supported by compilers (especially msvc). Hence we stick to gnu89 standard for now. Having said that, you can use c11's atomic as a **one of** implementation choices for the atomic_slist if you want, as long as there is alternative implementation for people who don't have c11.
